pkg/catalogv2/git/git.go

```diff
@@ -1,480 +1,440 @@
+	listOpts          *gogit.ListOptions
+	resetOpts         *gogit.ResetOptions
 }
 
-// LsRemote runs ls-remote on git repo and returns the HEAD commit SHA
-func (g *git) LsRemote(branch string, commit string) (string, error) {
-	if changed, err := g.remoteSHAChanged(branch, commit); err != nil || !changed {
-		return commit, err
+// BuildRepoConfig constructs and returns a new repository object for the given repository.
+// It requires a secret for authentication, the namespace, name of the repository,
+// the gitURL, a flag indicating if TLS verification should be skipped, and a CA bundle for SSL.
+// If the Git URL uses the SSH protocol, it checks if the URL is a valid SSH URL.
+// If the Git URL uses HTTP or HTTPS, it parses and verifies the URL.
+// It then constructs a directory path for the git repository.
+// It also sets an X-Install-Uuid header if an installation UUID is available.
+// If a CA bundle is provided, it converts the CA bundle from DER to PEM format, since Git requires PEM format.
+// In this case, insecureSkipTLS is set to false since a CA bundle is provided for secure communication.
+// Finally, it returns a new git object configured with these settings,
+// or an error if any step in this process fails.
+func BuildRepoConfig(secret *corev1.Secret, namespace, name, gitURL string, insecureSkipTLS bool, caBundle []byte) (*repository, error) {
+	isGitSSH, err := isGitSSH(gitURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify the type of URL %s: %w", gitURL, err)
 	}
-
-	output := &bytes.Buffer{}
-	if err := g.gitCmd(output, "ls-remote", "--", g.URL, formatRefForBranch(branch)); err != nil {
-		return "", err
+	if !isGitSSH {
+		u, err := url.Parse(gitURL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse URL %s: %w", gitURL, err)
+		}
+		if u.Scheme != "http" && u.Scheme != "https" {
+			return nil, fmt.Errorf("invalid git URL scheme %s, only http(s) supported", u.Scheme)
+		}
 	}
-
-	var lines []string
-	s := bufio.NewScanner(output)
-	for s.Scan() {
-		lines = append(lines, s.Text())
+	dir := gitDir(namespace, name, gitURL)
+	headers := map[string]string{}
+	if settings.InstallUUID.Get() != "" {
+		headers["X-Install-Uuid"] = settings.InstallUUID.Get()
+	}
+	// convert caBundle to PEM format because git requires correct line breaks, header and footer.
+	if len(caBundle) > 0 {
+		caBundle = convertDERToPEM(caBundle)
+		insecureSkipTLS = false
+	}
+
+	repo := &repository{
+		URL:               gitURL,
+		Directory:         dir,
+		caBundle:          caBundle,
+		insecureTLSVerify: insecureSkipTLS,
+		secret:            secret,
+		headers:           headers,
+		repoGogit:         &gogit.Repository{},
+		cloneOpts:         &gogit.CloneOptions{},
+		checkoutOpts:      &gogit.CheckoutOptions{},
+		fetchOpts:         &gogit.FetchOptions{},
+		listOpts:          &gogit.ListOptions{},
+		resetOpts:         &gogit.ResetOptions{},
+	}
+
+	// credentials must be setted before options
```

Comment:
nit: credentials must be set before options

Reply:
I have implemented a fix, please resolve the conversation